File: media-tools/compression/video_compression.py
from moviepy import VideoFileClip
import subprocess

# ANSI colors
CYAN = "\033[96m"
YELLOW = "\033[93m"
GREEN = "\033[92m"
RED = "\033[91m"
BLUE = "\033[94m"
MAGENTA = "\033[95m"
RESET = "\033[0m"

def get_video_bitrate(input_path):
    """Get bitrate of the input video in kbps using ffprobe"""
    try:
        result = subprocess.run(
            ['ffprobe', '-v', 'error', '-select_streams', 'v:0',
             '-show_entries', 'format=bit_rate', '-of', 'default=noprint_wrappers=1:nokey=1',
             input_path],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True
        )
        bitrate_bps = result.stdout.strip()
        if bitrate_bps.isdigit():
            return int(bitrate_bps) // 1000  # convert to kbps
        else:
            return None
    except Exception as e:
        print(f"{RED}Error getting bitrate:{RESET} {e}")
        return None

def compress_video(input_path, output_path, target_bitrate, target_height, preset='medium', threads=4):
    print(f"{CYAN}Starting compression...{RESET}")
    clip = VideoFileClip(input_path)
    clip_resized = clip.resized(height=target_height)
    clip_resized.write_videofile(
        output_path,
        bitrate=target_bitrate,
        codec='libx264',
        audio_codec='aac',
        preset=preset,
        threads=threads
    )
    clip.close()
    print(f"\n{GREEN}✅ Compressed video saved as: {output_path}{RESET}")

if __name__ == "__main__":
    input_path = input(f"{BLUE}Enter path to your video file:{RESET} ").strip().strip("'\"")
    default_output = input_path.rsplit('.', 1)[0] + "_compressed.mp4"
    output_path = input(f"{BLUE}Enter output path for compressed video [{default_output}]:{RESET} ").strip().strip("'\"")
    if not output_path:
        output_path = default_output

    orig_bitrate = get_video_bitrate(input_path)
    if orig_bitrate:
        print(f"{GREEN}Original video bitrate: {orig_bitrate} kbps{RESET}")
    else:
        print(f"{YELLOW}Could not determine original bitrate, defaulting to manual input.{RESET}")

    print(f"""
{MAGENTA}Choose compression option:{RESET}
1. Low quality (480p, ~500k bitrate)
2. Medium quality (720p, ~1500k bitrate)
3. High quality (1080p, ~3000k bitrate)
4. Custom: percentage of original bitrate (e.g. 50 for 50%)
""")

    choice = input(f"{BLUE}Enter option (1-4):{RESET} ").strip()

    if choice == "1":
        target_height = 480
        target_bitrate = "500k"
    elif choice == "2":
        target_height = 720
        target_bitrate = "1500k"
    elif choice == "3":
        target_height = 1080
        target_bitrate = "3000k"
    elif choice == "4":
        if orig_bitrate is None:
            print(f"{YELLOW}Original bitrate unknown. Please enter target bitrate manually (e.g., 1500k):{RESET}")
            target_bitrate = input(f"{BLUE}Target bitrate:{RESET} ").strip()
            target_height_str = input(f"{BLUE}Target height in pixels (e.g. 480, 720, 1080) [default 720]:{RESET} ").strip()
            target_height = int(target_height_str) if target_height_str else 720
        else:
            percent_str = input(f"{BLUE}Enter compression percentage (e.g. 50 for 50% of original bitrate):{RESET} ").strip()
            try:
                percent = float(percent_str)
                target_bitrate_value = int(orig_bitrate * (percent / 100))
                target_bitrate = f"{target_bitrate_value}k"
                target_height_str = input(f"{BLUE}Target height in pixels (e.g. 480, 720, 1080) [default 720]:{RESET} ").strip()
                target_height = int(target_height_str) if target_height_str else 720
            except Exception:
                print(f"{YELLOW}Invalid percentage entered, defaulting to 50% and 720p{RESET}")
                target_bitrate = f"{orig_bitrate // 2}k"
                target_height = 720
    else:
        print(f"{YELLOW}Invalid option, defaulting to Medium quality{RESET}")
        target_height = 720
        target_bitrate = "1500k"

    preset = input(f"{BLUE}Enter encoding preset (ultrafast, superfast, veryfast, faster, fast, medium, slow, slower, veryslow) [default medium]:{RESET} ").strip()
    if not preset:
        preset = "medium"

    threads_str = input(f"{BLUE}Enter number of CPU threads to use [default 4]:{RESET} ").strip()
    threads = int(threads_str) if threads_str.isdigit() else 4

    compress_video(input_path, output_path, target_bitrate, target_height, preset, threads)

File: media-tools/compression/video_compression.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include <cstdio>

using namespace std;

string get_video_bitrate(const string &input_path)
{
    string cmd =
        "ffprobe -v error -select_streams v:0 -show_entries format=bit_rate "
        "-of default=noprint_wrappers=1:nokey=1 \"" +
        input_path + "\"";

    FILE *pipe = popen(cmd.c_str(), "r");
    if (!pipe)
        return "";

    char buffer[128];
    string result = "";

    while (fgets(buffer, sizeof(buffer), pipe) != nullptr)
    {
        result += buffer;
    }
    pclose(pipe);

    // Remove whitespace
    while (!result.empty() && isspace(result.back()))
        result.pop_back();

    return result;
}

int main()
{
    string input_path;
    cout << "Enter path to your video file: ";
    getline(cin, input_path);

    string default_output = input_path.substr(0, input_path.find_last_of('.')) + "_compressed.mp4";

    string output_path;
    cout << "Enter output path [" << default_output << "]: ";
    getline(cin, output_path);
    if (output_path.empty())
        output_path = default_output;

    string orig_bitrate_str = get_video_bitrate(input_path);
    int orig_bitrate_kbps = 0;

    if (!orig_bitrate_str.empty())
    {
        orig_bitrate_kbps = stoi(orig_bitrate_str) / 1000;
        cout << "Original bitrate: " << orig_bitrate_kbps << " kbps\n";
    }
    else
    {
        cout << "Could not determine bitrate.\n";
    }

    cout << "\nChoose compression option:\n"
         << "1. Low (480p, 500k)\n"
         << "2. Medium (720p, 1500k)\n"
         << "3. High (1080p, 3000k)\n"
         << "4. Custom percentage of original bitrate\n"
         << "Enter option (1-4): ";

    string choice;
    getline(cin, choice);

    int target_height = 720;
    string target_bitrate = "1500k";

    if (choice == "1")
    {
        target_height = 480;
        target_bitrate = "500k";
    }
    else if (choice == "2")
    {
        target_height = 720;
        target_bitrate = "1500k";
    }
    else if (choice == "3")
    {
        target_height = 1080;
        target_bitrate = "3000k";
    }
    else if (choice == "4")
    {
        if (orig_bitrate_kbps == 0)
        {
            cout << "Enter target bitrate (e.g., 1500k): ";
            getline(cin, target_bitrate);
        }
        else
        {
            cout << "Enter percentage (e.g., 50 for 50%): ";
            string percent_str;
            getline(cin, percent_str);
            int percent = stoi(percent_str);

            int target_kbps = orig_bitrate_kbps * percent / 100;
            target_bitrate = to_string(target_kbps) + "k";
        }

        cout << "Enter target height (default 720): ";
        string height_str;
        getline(cin, height_str);
        if (!height_str.empty())
            target_height = stoi(height_str);
    }
    else
    {
        cout << "Invalid option. Using medium.\n";
    }

    string preset;
    cout << "Enter preset (default medium): ";
    getline(cin, preset);
    if (preset.empty())
        preset = "medium";

    cout << "Threads [default 4]: ";
    string threads_str;
    getline(cin, threads_str);
    int threads = threads_str.empty() ? 4 : stoi(threads_str);

    // Build FFmpeg command
    string cmd =
        "ffmpeg -i \"" + input_path + "\" "
                                      "-vf scale=-2:" +
        to_string(target_height) + " "
                                   "-b:v " +
        target_bitrate + " "
                         "-preset " +
        preset + " "
                 "-threads " +
        to_string(threads) + " "
                             "-c:v libx264 -c:a aac "
                             "\"" +
        output_path + "\"";

    cout << "\nRunning:\n"
         << cmd << "\n\n";

    system(cmd.c_str());

    cout << "\nDone! Output saved to " << output_path << endl;

    return 0;
}
